shell/plugins/codemirror.js

```diff
@@ -167,3 +167,157 @@ CodeMirror.registerHelper('fold', 'yamlcomments', (cm, start) => {
     };
   }
 });
+
+/**
+ * It display a dot for each space character in the text;
+ * used in combination with 'cm-markdown-line-breaks' in CodeMirror.vue to display line break markdowns
+ */
+CodeMirror.defineOption('showMarkdownLineBreaks', false, (codeMirror) => {
+  codeMirror.addOverlay({
+    name:  'show-markdown-line-breaks',
+    token: (stream) => {
+      if (stream.string[stream.pos].match(/\s/)) {
+        stream.next();
+
+        return stream.pos % 2 === 0 ? 'markdown-single-trailing-space-even' : 'markdown-single-trailing-space-odd';
+      }
+
+      stream.next();
+
+      return null;
+    }
+  });
+});
+
+/**
+ * It enables the text color selection in CodeMirror.vue
+ */
+CodeMirror.defineOption('styleSelectedText', false, (cm, val, old) => {
+  const prev = old && old !== CodeMirror.Init;
+
+  if (val && !prev) {
+    cm.state.markedSelection = [];
+    cm.state.markedSelectionStyle = typeof val === 'string' ? val : 'CodeMirror-selectedtext';
+    reset(cm);
+    cm.on('cursorActivity', onCursorActivity);
+    cm.on('change', onChange);
+  } else if (!val && prev) {
+    cm.off('cursorActivity', onCursorActivity);
+    cm.off('change', onChange);
+    clear(cm);
+    cm.state.markedSelection = cm.state.markedSelectionStyle = null;
+  }
+});
+
+function onCursorActivity(cm) {
+  if (cm.state.markedSelection) {
+    cm.operation(() => {
+      update(cm);
+    });
+  }
+}
+
+function onChange(cm) {
+  if (cm.state.markedSelection && cm.state.markedSelection.length) {
+    cm.operation(() => {
+      clear(cm);
+    });
+  }
+}
+
+const CHUNK_SIZE = 8;
+const Pos = CodeMirror.Pos;
+const cmp = CodeMirror.cmpPos;
+
+function coverRange(cm, from, to, addAt) {
+  if (cmp(from, to) === 0) {
+    return;
+  }
+  const array = cm.state.markedSelection;
+  const cls = cm.state.markedSelectionStyle;
+
+  for (let line = from.line;;) {
+    const start = line === from.line ? from : Pos(line, 0);
+    const endLine = line + CHUNK_SIZE; const atEnd = endLine >= to.line;
+    const end = atEnd ? to : Pos(endLine, 0);
+    const mark = cm.markText(start, end, { className: cls });
+
+    if (addAt === null || addAt === undefined) {
+      array.push(mark);
+    } else {
+      array.splice(addAt++, 0, mark);
+    }
+    if (atEnd) {
+      break;
+    }
+    line = endLine;
+  }
+}
+
+function clear(cm) {
```

Comment:
See comment above about CodeMirror add-ons.

Reply:
What does it mean? I was asking what's the scope of the function, while the comment on the top states just this:

```
/*
 * NOTE: This isn't actually a real plugin anymore, it's is dynamically loaded in components/CodeMirror.vue
 * so that it doesn't all get loaded put into vendor.js
 */
```

Reply:
Ok it was the wrong comment 😅 
https://github.com/rancher/dashboard/blob/f263c4ffbc2e63ab7703d55f5a6b438ad65ede92/shell/plugins/codemirror.js#L193-L198